### PR TITLE
UCT:  API interface changes for supporting memory type detection

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

File: src/ucp/api/ucp_def.h [hunk 1]
```diff
@@ -141,6 +141,13 @@ typedef struct ucp_rkey                  *ucp_rkey_h;
  */
 typedef struct ucp_mem                   *ucp_mem_h;
 
+/*
+ * @ingroup UCP_MEM
+ * @brief UCP memory type
+ *
+ * Memory type handle is an opaque object representing a memory adreess type
```
Comment (miked-mellanox) [inline src/ucp/api/ucp_def.h]: s/adreess/address/

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4795/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4794/ for details (Mellanox internal link).

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -385,6 +385,16 @@ enum {
                                                sockaddr */
 };
 
+/*
+ * @ingroup UCT_MD
+ * @brief  Memory types
+ */
+typedef enum {
+    UCT_MD_MEM_TYPE_DEFAULT = 0,   /**< Default system memory */
+    UCT_MD_MEM_TYPE_CUDA,          /**< NVIDIA CUDA memory */
```
Comment (shamisp) [inline src/uct/api/uct.h]: s/CUDA/GPU ?

I don't think it is a good idea to go with a very specific naming.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4806/ for details (Mellanox internal link).

File: src/uct/api/uct.h (hunk 2 above)
Comment (bureddy) [reply src/uct/api/uct.h]: @shamisp the reason for this to differentiate multiple GPU Types (nvidia, amd, ..etc)

Comment (shamisp) [reply src/uct/api/uct.h]: I don't like the usage of vendor specific naming in the API. It supposed to be abstracted in some way.

Comment (bureddy) [reply src/uct/api/uct.h]: @shamisp I can change this to s/CUDA/GPU given that there may not be heterogeneous GPU support.  one another way is to enumerate (GPU_TYPE_0, GPY_TYPE_1..).

Comment (shamisp) [reply src/uct/api/uct.h]: Yes, we can create more than one type of GPUs. @yosefe your take ?

Comment (yosefe) [reply src/uct/api/uct.h]: i think it should be "CUDA", since the user would pass to this API memory buffers allocated by cuda library. which should be distinguished from passing pointers allocated by rocm library. 
So the reason to have this is because user application already has to be aware of the memory type.

File: src/ucp/api/ucp_def.h [hunk 3]
```diff
@@ -141,6 +141,13 @@ typedef struct ucp_rkey                  *ucp_rkey_h;
  */
 typedef struct ucp_mem                   *ucp_mem_h;
 
+/*
+ * @ingroup UCP_MEM
+ * @brief UCP memory type
+ *
+ * Memory type handle is an opaque object representing a memory address type
+*/
+typedef struct ucp_mem_type              *ucp_mem_type_h;
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: why need a handle instead of just enum? and why do we need this in the UCP API?

File: src/ucp/core/ucp_mm.c [hunk 4]
```diff
@@ -106,6 +112,36 @@ static ucs_status_t ucp_memh_reg_mds(ucp_context_h context, ucp_mem_h memh,
     return UCS_OK;
 }
 
+ucs_status_t ucp_addr_domain_detect_mds(ucp_context_h context, void *addr, ucp_mem_type_h mem_type_h)
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c]: pls add unit test for this funciton

File: src/ucp/core/ucp_mm.c [hunk 5]
```diff
@@ -106,6 +112,36 @@ static ucs_status_t ucp_memh_reg_mds(ucp_context_h context, ucp_mem_h memh,
...
+    unsigned md_index;
+    uct_memory_type_t mem_type = UCT_MD_MEM_TYPE_DEFAULT;
+
+    mem_type_h->md_map = 0;
+    mem_type_h->id = UCT_MD_MEM_TYPE_DEFAULT;
+
+    /*TODO: return if no MDs with address domain detect */
+
+    for (md_index = 0; md_index < context->num_mds; ++md_index) {
+        if (context->tl_mds[md_index].attr.cap.mem_type != UCT_MD_MEM_TYPE_DEFAULT) {
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c]: pls add a shortlist of md's whose memory types have to be checked instead of going over all.
potentially, the implementation of this function+test has to go in a separate PR, and keep this PR for UCT API only.

File: src/uct/api/uct.h [hunk 6]
```diff
@@ -385,6 +385,16 @@ enum {
...
+    UCT_MD_MEM_TYPE_LAST    = UCT_MD_MEM_TYPE_CUDA
```
Comment (yosefe) [inline src/uct/api/uct.h]: "LAST" should be right after CUDA (which would generate CUDA+1) rather than =CUDA
this way it would be possible to create arrays which entry for each memory type e.g 
`size_t max_zcopy[UCT_MD_MEM_TYPE_LAST]`

File: src/uct/api/uct.h (hunk 2 above)
Comment (shamisp) [reply src/uct/api/uct.h]: It the enum is expected to be exposed in UCP ?

Comment (bureddy) [reply src/uct/api/uct.h]: @shamisp  we may need to know no.of  types.   but not the enum values directly.

Comment (yosefe) [reply src/uct/api/uct.h]: @shamisp i think we should, at least as flags for ucp_init() - which memory types should UCX support. This way applications/libraries can enable/disable support for GPU buffers

**Comment by bureddy** [comment]
@yosefe kept only UCT/API changes in this PR. will push another PR for UCP func+unittest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4836/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4838/ for details (Mellanox internal link).

File: src/uct/sm/cma/cma_md.c [hunk 7]
```diff
@@ -59,7 +59,8 @@ static ucs_status_t uct_cma_md_open(const char *md_name, const uct_md_config_t *
         .mem_free     = (void*)ucs_empty_function_return_success,
         .mkey_pack    = (void*)ucs_empty_function_return_success,
         .mem_reg      = uct_cma_mem_reg,
-        .mem_dereg    = (void*)ucs_empty_function_return_success
+        .mem_dereg    = (void*)ucs_empty_function_return_success,
+        .mem_type_detect   = ucs_empty_function_return_unsupported,
```
Comment (shamisp) [inline src/uct/sm/cma/cma_md.c]: why we don't return UCT_MD_MEM_TYPE_DEFAULT for all the rest non-cuda devices ?

File: src/uct/api/uct.h (hunk 2 above)
Comment (shamisp) [reply src/uct/api/uct.h]: I think, I can leave with CUDA memory at UCT layer, but I don't think it has to be exposed as "CUDA" in UCP layer.
And yes, we have to add something to init to give a hint about GPU usage.

File: src/uct/sm/cma/cma_md.c (hunk 7 above)
Comment (bureddy) [reply src/uct/sm/cma/cma_md.c]: ucp function will call mem detect only if md supports non default mem type.  I think we can't simply return default for any address without detection logic.

File: src/uct/api/uct.h (hunk 2 above)
Comment (khamidouche) [reply src/uct/api/uct.h]: I agree with @shamisp that it will be better if we can avoid the types in UCP. but at the same time I'm wondering how UCP can select which UCT to use if it does not know the exact type. Because I can have CUDA installed in ROCM node which means I can configure UCX with both ROCM and CUDA support

File: src/uct/sm/cma/cma_md.c (hunk 7 above)
Comment (shamisp) [reply src/uct/sm/cma/cma_md.c]: so, you want to have the branch on the UCP level, right ?

File: src/uct/api/uct.h (hunk 2 above)
Comment (shamisp) [reply src/uct/api/uct.h]: @khamidouche agreed. I'm not sure how the detection will work. It is reasonable to assume that some system will have both types of GPUs.

File: src/uct/sm/cma/cma_md.c (hunk 7 above)
Comment (bureddy) [reply src/uct/sm/cma/cma_md.c]: yes

Comment (shamisp) [reply src/uct/sm/cma/cma_md.c]: hmmm ... how ? You need some sort of capabilities flag for this.

Comment (bureddy) [reply src/uct/sm/cma/cma_md.c]: this PR also adding mem  type in  md_attr (md_attr->cap.mem_type). So, will check this flag.
UCP changes are here: https://github.com/openucx/ucx/pull/1923

Comment (shamisp) [reply src/uct/sm/cma/cma_md.c]: okay.

File: src/uct/api/uct.h [hunk 8]
```diff
@@ -633,6 +643,7 @@ struct uct_md_attr {
         size_t               max_alloc; /**< Maximal allocation size */
         size_t               max_reg;   /**< Maximal registration size */
         uint64_t             flags;     /**< UCT_MD_FLAG_xx */
+        uct_memory_type_t    mem_type;  /**< Supported memory type */
```
Comment (yosefe) [inline src/uct/api/uct.h]: How do we express the fact IB supports CUDA memory type (e.g peer-direct)?

**Comment by bureddy** [comment]
@yosefe yes. we need  mem type register capability to express peer-direct.

**Comment by bureddy** [comment]
MEM_TYPE_DEFAULT ==> MEM_TYPE_HOST

**Comment by bureddy** [comment]
made following changes based on the todays discussion on the call

   - s/UCT_MD_MEM_TYPE_DEFAULT/UCT_MD_MEM_TYPE_HOST
   -  API name : s/uct_md_mem_type_detect/uct_md_is_mem_type_owned
   -  added length param in uct_md_is_mem_type_owned api for more portability
   -  mem type registration capability flag (mem_type_reg_flags)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4856/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
adding @bbenton

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4887/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4898/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@yosefe please review.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4938/ for details (Mellanox internal link).

File: src/uct/api/uct.h [hunk 9]
```diff
@@ -396,6 +396,17 @@ typedef enum {
...
 
 
+/*
+ * @ingroup UCT_MD
+ * @brief  Memory type owned status
+ */
+typedef enum {
+    UCT_MD_MEM_TYPE_OWNED = 0,          /**< MD owns the memory type */
+    UCT_MD_MEM_TYPE_NOT_OWNED,          /**< MD not owns the memort type */
+    UCT_MD_MEM_TYPE_NOT_DETECTABLE      /**< MD can't detect memory owned or not */
```
Comment (yosefe) [inline src/uct/api/uct.h]: why do we need that?

Comment (bureddy) [reply src/uct/api/uct.h]: it seems semantically it is not clean to return NOT_OWNED for the HOST memory  on  non-cuda MDs.

Comment (yosefe) [reply src/uct/api/uct.h]: an MD should return either "true" (definitely owned) or "false" (definitely not owned). "false" does not necessarily mean it's host memory; however, if all memory domains returned "not owned" UCP would conclude this is host memory. This conclusion would also be correct assuming we have support in UCX for all possible memory types in the application.

Comment (bureddy) [reply src/uct/api/uct.h]: we can return "true" or "false" if we conclude that it is host memory if all MD return "no owned"
will update PR

Comment (yosefe) [reply src/uct/api/uct.h]: BTW, when i said "true"/"false" i didn't mean that we need to define such constants, just plain old 0/1

Comment (bureddy) [reply src/uct/api/uct.h]: updated PR.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4947/ for details (Mellanox internal link).

File: src/uct/base/uct_md.c [hunk 10]
```diff
@@ -544,15 +544,12 @@ int uct_md_is_sockaddr_accessible(uct_md_h md, const ucs_sock_addr_t *sockaddr,
...
-                                                        void *addr, size_t length)
+int uct_md_is_mem_type_owned(uct_md_h md, void *addr, size_t length)
 {
     return md->ops->is_mem_type_owned(md, addr, length);
 }
 
-uct_memory_type_owned_status_t uct_md_return_mem_type_not_detectable(uct_md_h md,
-                                                                        void *addr,
-                                                                        size_t length)
+int uct_md_return_mem_type_not_owned(uct_md_h md, void *addr, size_t length)
```
Comment (yosefe) [inline src/uct/base/uct_md.c]: no need to define it, can use ucs_empty_function_return_zero instead

Comment (bureddy) [reply src/uct/base/uct_md.c]: I defined this to avoid incompatible assignment warning because ucs_empty_function_return_zero return type is unsigned. 
I get way with warning with (void *) type casting. Updated PR.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4953/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@bbenton can you please take a look ?

**Comment by shamisp** [comment]
👍

**Comment by gmegan** [comment]
+1
